### ISSUE TITLE
Storage collateral vs pledge collateral

### ISSUE THREAD

**Issue by anorth** [filecoin-project/specs#294]
The breach of contract dispute in [faults](https://github.com/filecoin-project/specs/blob/master/faults.md) states that the penalty for defaulting on a contract is taken from pledge collateral. It's not explicit about whether the fee is paid to the storage client or burned.

In the actor spec, [ArbitrateDeal](https://github.com/filecoin-project/specs/blob/master/actors.md#arbitratedeal) pseudocode mentions both burning pledge collateral and transferring storage collateral to the client. How this storage collateral is represented and where it comes from isn't specified.

[On being slashed](https://github.com/filecoin-project/specs/blob/master/mining.md#on-being-slashed-wip-needs-discussion) in mining spec opens a discussion and TODO for disambiguating the two collaterals throughout the spec. My impression from conversations elsewhere suggest we're currently intending to identify both.

If we do follow through with storage collateral, we'll need:
- a mechanism for determining how much storage collateral is to be posted (requested by client? network constant?)
- a mechanism for posting that collateral
- a mechanism for retrieving or repurposing the collateral when a deal expires
- `ArbitrateDeal` to transfer some of that to the client when a dispute is successful

Accounting for storage collateral is complicated somewhat by the fact that the details of a successful deal are never posted on chain.

**Comment by anorth** [comment]
One avenue that I think is worth investigating is avoiding having the two collaterals, and giving a client some rights to a miner's pledge collateral in the event of a breach of contract or storage fault, instead of burning it (or recovering it from the flames after the fact). If a miner isn't bust, they may become under-collateralised (which can happen in other ways too), and we'll have a standard mechanism for them to address the situation.

Mechanisms for a client to prove entitlement to the pledge collateral are probably no more complex than those used to prove entitlement to a distinguished storage collateral.

The largest downside I see is that it caps the collateral that can be associated with a deal; it's conceivable that a client could desire more.

**Comment by pooja** [comment]
cc @decentralion as this may relate to your collateral spec work

**Comment by pooja** [comment]
Also related to #60

**Comment by decentralion** [comment]
tl;dr: We should adopt @anorth's suggestion and drop storage collateral from the spec.

---

From a conceptual vantage point, I see pledge and storage collateral as quite different, which corresponds to seeing maintaining consensus and storing data as different roles. Miners play a dual role as keepers of consensus (compensated by block rewards, secured by pledge collateral) and storers of user data (compensated by storage fees, secured by storage collateral). In the event that a miner does a bad job of securing consensus (e.g. mining on multiple forks) they forfeit collateral to the network as a whole (by burning it). If they do a bad job of storing user data, then they forfeit collateral to the affected party (by paying it back to the user).

Framing things this way gets us some desirable properties:
- the storage collateral amount is set by the user, who actually knows how valuable the data is to them, so they can choose a point on the risk/return continuum that best suits them
- we might even have a future with "storage-only" miners, who don't participate in consensus, but only store data. this would have a cheaper cost structure (don't need pledge collateral, maybe don't even need a full node?) and would allow filecoin to offer "extremely cheap but low reliability" storage if there's a market for that.

However, getting storage collateral right is pretty complicated. First of all, we wouldn't want storage collateral to be priced in FIL, because the value of the users data doesn't fluctuate with the filecoin exchange rate. One of the big risks that storage users will face is the risk that the FIL price crashes, which causes storage rates (as denominated in FIL) to skyrocket, which causes miners to throw away existing users' data in search of new contracts. However, FIL-denominated storage collateral cannot mitigate this risk, because the FIL price crash would also cause the value of the collateral to crash. To do this properly, we would need support for other currencies as storage collateral.

Also, repair miners will have a huge role to play in storage collateral. Let's suppose I value my data at $1000/GB. Rather than finding an individual miner with amazingly robust storage guarantees, it would make more sense to find a number of uncorrelated miners that individually have much lower guarantees, and then have a repair miner orchestrate across them.

Also, storage collateral probably shouldn't be handled via on-chain deposits or locked up capital. Think of it like insurance: I get a contract with a repair miner who, in addition to storing my data, offers me an insurance policy paying $10k if they lose my 10GB of data. Actually keeping $10k locked up on chain would incur capital costs of (say) $1k/year, which is massively too expensive and would render the market uneconomical. So, like real-world insurance operations, they would manage policy risk and keep relatively small reserves to cover those policies that actually do pay out (but with a risk of going bankrupt, and corresponding needs for reputation / trust / regulation). 

Putting this all together: to accomplish the aims of storage collateral properly is actually really complicated, and not in scope for the work we're doing between now and launch. So, we have two options:

- Implement storage collateral right now in a very simplistic way (e.g. user specifies amount of FIL/GB that the miner must provide, and that amount gets sent to user if the contract is breached). This will actually enormously complicate the market (price is now 2-dimensional) while also being too expensive to actually let users price the value of their data, and won't protect against FIL price volatility. 
- Punt on storage collateral for now, and possibly make a fraction of burned pledge collateral go to users' in the event of a failure (per @anorth's [suggestion](https://github.com/filecoin-project/specs/issues/294#issuecomment-496084913))

Of these, the second is clearly better. I'm not sure about plans for smart contract support in Filecoin, but I'll note that the simplistic storage collateral described above could be implemented via smart contracting, rather than additions at the protocol layer. Better yet, the more sophisticated insurance models that would actually provide good economics could also be implemented via smart contracts. It would be ideal if Filecoin events are published and consumable on other protocols that are purpose-built for enabling sophisticated financial agreements. This will enable decentralized innovation around file storage insurance models, in an environment with good support and access to stablecoins, and we won't have to maintain it. :)

**Comment by whyrusleeping** [comment]
Doing anything with pledge collateral other than burning it defeats much of the purpose of pledge collateral (a malicious miner will be their own client and simply use this to make their attacks cheaper, thus necessitating a proportionally higher pledge collateral for the same level of security). 

On the topic of storage collateral, I'm indifferent. I agree with @decentralion's points about the cost structure not making sense, but also think that clients should be able to be compensated if their data is lost (and that can't come from pledge collateral for previously stated reasons). That said, many hours have been spent arguing this point back and forth, and the last one of those ended with agreement that we would have some simple storage collateral. If we want to change that, i'm gonna stay out of it and let y'all work it through with @jbenet

**Comment by anorth** [comment]
@pooja, thanks, I didn't find #60, sorry for splitting the discussion. I now realise this has opened up a larger discussion that was already previously resolved in favour of designing a storage collateral model. I'm going to close this issue and take immediate parts to #60.

**Comment by decentralion** [comment]
> Doing anything with pledge collateral other than burning it defeats much of the purpose of pledge collateral (a malicious miner will be their own client and simply use this to make their attacks cheaper, thus necessitating a proportionally higher pledge collateral for the same level of security).

I was imagining we might have a construction where some proportion `p` gets returned to the user whose data was dropped, and `1-p` is burned. But you're right that this would allow a malicious miner to "withdraw" `p` of their pledge collateral back to their own account, and then start launching attacks without the fear of slashing-based disincentives. So, any transfers out of pledge collateral significantly change the safety guarantees we get from it.

**Comment by huaxiaone** [comment]
If there's a smart contract on this chain, there's no need to discuss these details. You're great @decentralion
